File: agent/modules/shelters.js
import {SERVICE_URL} from './config.js';
import {fetchJSON,getJSON} from './client.js';
export const SHELTER_TYPES={earthquake:['地震','地震',4],tsunami:['津波','海啸',5],landslide:['土砂','滑坡泥石流',2],flood:['洪水','洪水',1],'storm-surge':['高潮','风暴潮',3],'inland-flood':['内水氾濫','内涝',7],volcano:['火山','火山',8]};
export function meters(lat,lon,a,b){const r=Math.PI/180,p=lat*r,q=a*r;return 6371000*2*Math.asin(Math.min(1,Math.sqrt(Math.sin((q-p)/2)**2+Math.cos(p)*Math.cos(q)*Math.sin((b-lon)*r/2)**2)));}
export function selectShelters(collections,p,hazard){
 const type=SHELTER_TYPES[hazard];if(!type)throw new Error('choose-disaster');const found=new Map();
 for(const collection of collections)for(const f of collection.features||[]){
  const g=f.geometry,props=f.properties||{};if(g?.type!=='Point'||![1,'1'].includes(props['disaster'+type[2]]))continue;
  const [longitude,latitude]=g.coordinates||[];if(!Number.isFinite(latitude)||!Number.isFinite(longitude)||latitude<20||latitude>46.5||longitude<122||longitude>154||!props.name)continue;
  const distanceMeters=Math.round(meters(p.latitude,p.longitude,latitude,longitude));if(distanceMeters>10000)continue;
  const id=latitude+','+longitude+':'+props.name;
  found.set(id,{id,name:String(props.name),address:String(props.address||''),remarks:String(props.remarks||''),latitude,longitude,distanceMeters,hazard,designated:true,openingStatus:'unverified',routeSafety:'unverified',elevationMeters:null,source:'https://www.gsi.go.jp/bousaichiri/hinanbasho'});
 }
 return [...found.values()].sort((a,b)=>a.distanceMeters-b.distanceMeters).slice(0,8);
}
export function navigationURL(place){if(!place?.designated||!SHELTER_TYPES[place.hazard]||!Number.isFinite(place.latitude)||!Number.isFinite(place.longitude))throw new Error('invalid-destination');return 'https://www.google.com/maps/dir/?api=1&destination='+place.latitude+','+place.longitude+'&travelmode=walking';}
export async function nearbyShelters(position,hazard,sameOrigin=false){
 if(!SHELTER_TYPES[hazard])throw new Error('choose-disaster');
 if(!Number.isFinite(position.latitude)||!Number.isFinite(position.longitude)||position.accuracy>1000)throw new Error('location-inaccurate');
 const query='/api/shelters?lat='+position.latitude.toFixed(5)+'&lon='+position.longitude.toFixed(5)+'&hazard='+encodeURIComponent(hazard);
 if(sameOrigin)return fetchJSON(query);
 if(SERVICE_URL)return getJSON(query);
 const z=10,n=2**z,x=Math.floor((position.longitude+180)/360*n),y=Math.floor((1-Math.asinh(Math.tan(position.latitude*Math.PI/180))/Math.PI)/2*n);
 const jobs=[];for(const dx of [-1,0,1])for(const dy of [-1,0,1])jobs.push(fetchJSON('https://cyberjapandata.gsi.go.jp/xyz/skhb'+String(SHELTER_TYPES[hazard][2]).padStart(2,'0')+'/'+z+'/'+(x+dx)+'/'+(y+dy)+'.geojson',12000));
 const results=await Promise.allSettled(jobs),collections=results.filter(r=>r.status==='fulfilled').map(r=>r.value);
 const places=selectShelters(collections,position,hazard);
 if(hazard==='tsunami')await Promise.all(places.slice(0,3).map(async place=>{try{const data=await fetchJSON('https://cyberjapandata2.gsi.go.jp/general/dem/scripts/getelevation.php?lon='+place.longitude+'&lat='+place.latitude+'&outtype=JSON',8000);if(Number.isFinite(data.elevation))place.elevationMeters=data.elevation;}catch{}}));
 return {places,state:collections.length===9?'fetched':collections.length?'partial':'unavailable',checkedAt:Date.now(),source:'https://www.gsi.go.jp/bousaichiri/hinanbasho'};
}

File: server/shelters.py
"""GSI designated emergency evacuation places, filtered by disaster suitability."""
import concurrent.futures
import json
import math
import subprocess
import threading
import time

# GSI's layer IDs are skhb01..08; a different disaster never substitutes for a match.
LAYERS={'flood':1,'landslide':2,'storm-surge':3,'earthquake':4,'tsunami':5,'fire':6,'inland-flood':7,'volcano':8}
SOURCE='https://www.gsi.go.jp/bousaichiri/hinanbasho'
_cache={};_lock=threading.Lock()


def point(lat,lon):
    lat,lon=float(lat),float(lon)
    if not math.isfinite(lat) or not math.isfinite(lon) or not (20<=lat<=46.5 and 122<=lon<=154):raise ValueError('outside-japan')
    return lat,lon


def distance(lat,lon,a,b):
    p,q=map(math.radians,(lat,a));dp=q-p;dl=math.radians(b-lon)
    return 6371000*2*math.asin(min(1,math.sqrt(math.sin(dp/2)**2+math.cos(p)*math.cos(q)*math.sin(dl/2)**2)))


def tiles(lat,lon,z=10):
    x=int((lon+180)/360*2**z);y=int((1-math.asinh(math.tan(math.radians(lat)))/math.pi)/2*2**z)
    return [(x+dx,y+dy) for dx in (-1,0,1) for dy in (-1,0,1)]


def fetch_json(url):
    with _lock:
        cached=_cache.get(url)
        if cached and time.time()-cached[0]<3600:return cached[1]
    p=subprocess.run(['curl','-sS','--max-time','10','--max-filesize','3000000','-w','\n%{http_code}',url],capture_output=True,timeout=12)
    body,_,status=p.stdout.rpartition(b'\n')
    if p.returncode or status not in (b'200',b'404'):raise OSError('gsi-unavailable')
    data={'type':'FeatureCollection','features':[]} if status==b'404' else json.loads(body)
    with _lock:
        if len(_cache)>300:_cache.clear()
        _cache[url]=(time.time(),data)
    return data


def select_places(collections,lat,lon,hazard,radius=10000):
    layer=LAYERS.get(hazard)
    if not layer:raise ValueError('choose-specific-disaster')
    found={}
    for collection in collections:
        for feature in collection.get('features',[]):
            g=feature.get('geometry') or {};p=feature.get('properties') or {}
            if g.get('type')!='Point' or p.get('disaster'+str(layer)) not in (1,'1'):continue
            try:a,b=point(g['coordinates'][1],g['coordinates'][0])
            except (ValueError,KeyError,IndexError,TypeError):continue
            d=distance(lat,lon,a,b)
            if d>radius or not p.get('name'):continue
            key=f'{a:.6f},{b:.6f}:{p["name"]}'
            found[key]={'id':key,'name':str(p['name']),'address':str(p.get('address') or ''),'remarks':str(p.get('remarks') or ''),
                        'latitude':a,'longitude':b,'distanceMeters':round(d),'hazard':hazard,'designated':True,
                        'openingStatus':'unverified','routeSafety':'unverified','elevationMeters':None,
                        'source':SOURCE,'mapUrl':f'https://www.google.com/maps/dir/?api=1&destination={a},{b}&travelmode=walking'}
    return sorted(found.values(),key=lambda p:p['distanceMeters'])[:8]


def nearby(lat,lon,hazard,fetcher=fetch_json):
    lat,lon=point(lat,lon);layer=LAYERS.get(hazard)
    if not layer:raise ValueError('choose-specific-disaster')
    collections=[];errors=0
    urls=[f'https://cyberjapandata.gsi.go.jp/xyz/skhb{layer:02}/10/{x}/{y}.geojson' for x,y in tiles(lat,lon)]
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as pool:
        for future in concurrent.futures.as_completed([pool.submit(fetcher,u) for u in urls]):
            try:collections.append(future.result())
            except (OSError,ValueError,subprocess.SubprocessError):errors+=1
    places=select_places(collections,lat,lon,hazard)
    if hazard=='tsunami' and fetcher is fetch_json:
        def elevation(p):
            try:
                data=fetcher(f'https://cyberjapandata2.gsi.go.jp/general/dem/scripts/getelevation.php?lon={p["longitude"]}&lat={p["latitude"]}&outtype=JSON')
                value=data.get('elevation')
                if isinstance(value,(int,float)) and math.isfinite(value):p['elevationMeters']=round(value,1)
            except (OSError,ValueError,subprocess.SubprocessError):pass
        with concurrent.futures.ThreadPoolExecutor(max_workers=3) as pool:list(pool.map(elevation,places[:3]))
    return {'places':places,'hazard':hazard,'radiusMeters':10000,'source':SOURCE,
            'checkedAt':time.strftime('%Y-%m-%dT%H:%M:%SZ',time.gmtime()),'state':'unavailable' if errors==len(urls) else 'partial' if errors else 'fetched',
            'notice':'災害種別に対応した指定緊急避難場所。開設・通行可否は未確認 / 按灾种指定的紧急避难场所；开放及道路状况未核实'}
